File: rcpp_generic_mcmc/mcmc/R/gibbs.R
cp_substate <- function(state, substate) {
  out <- NULL

  for (s in substate) {
    out$s  <- state$s
  }

  out
}

gibbs <- function(state, update_fn, B, burn, print_freq=0, monitors=NULL, thin=NULL) {
  out <- rep(list(state), B)

  for (i in 1:burn) {
    update_fn(state)
  }

  for (i in 1:B) {
    update_fn(state)
    out[[i]] <- state
  }

  return(out)
}

File: cythonLearn/MCMC_pure_py/gibbs.py
import copy


def get_substate(state, monitor):
    out = dict()
    for k in monitor:
        out[k] = copy.deepcopy(state.__getattribute__(k))
    return out


def gibbs(init, update_fn, nmcmc: int, burn: int, monitors=None,
          printFreq: int=0):

    # Create state object
    state = copy.deepcopy(init)

    # Create monitor
    if monitors is None:
        monitors = [list(state.__dict__.keys())]

    # Preallocate memory for output
    out = [[get_substate(state, mtr) for mtr in monitors]
           for i in range(nmcmc)]

    # Print Message
    def printMsg(mcmc_iter: int):
        if printFreq is not 0 and mcmc_iter % printFreq == 0:
            print("{} / {}".format(mcmc_iter, nmcmc))

    # Burn
    for i in range(burn):
        printMsg(i)
        update_fn(state)

    # Gibbs
    for i in range(nmcmc):
        printMsg(i + burn)
        update_fn(state)

        # Grab params to put in output
        out[i] = [get_substate(state, mtr) for mtr in monitors]

    return out

# Implement the individual update functions for each parameter
# in update_fn using Cython.
